some more fixes

neheb opened the pull request:
I'm not sure about the shared_ptr change. Cyclic references are avoided by passing by value. I doubt there's much overhead here.

neheb committed:
clang-tidy: remove unused headers

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
clang-tidy: try to add missing includes

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
remove older compiler compat

No need to initialize the base class for copy constructors.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
add some const to casts

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb commented:
more false positives. unique_ptr is not a vector.

ping @kevinbackhouse

neheb committed:
codeql: do not warn to use at for unique_ptr

unique_ptr does not have at()

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
vector to unique_ptr

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb commented:
some more make_uniques. Let's see if it complains now.

neheb commented:
The last commit changes to unique_ptr as only a raw pointer is used. No vector functionality like push_back. Probably uses less memory.

As for the first, it looks like get() is just used to avoid having a second smart pointer. I don’t know if that’s a good idea.

neheb commented:
I'll rebase once that's merged.

neheb committed:
disable ICC test

Some platforms such as Fedora use zlib-ng instead of zlib, which has
different compressed output. Compressed output shouldn't be tested
anyway.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
reenable webp test

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
tests: remove itertools

No need for it.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
pylint: use \x

E2513: Invalid unescaped character esc, use "\x1B" instead. (invalid-character-esc)

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
pylint: avoid pointless string concat

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
remove m_ from class

It's only used in a single function. No need to have it in the class.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
remove unused using

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
fix warning on Cygwin

When calling std::vector::resize(0), Cygwin warns on it. Add a check to
avoid that.

warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
remove unnecessary commas

Signed-off-by: Rosen Penev <rosenp@gmail.com>

kevinbackhouse commented on src/preview.cpp:
```diff
@@ -819,8 +819,9 @@
-  decodeHexTable.fill(invalid);
+  auto decodeHexTable = std::make_unique<byte[]>(256);
+  for (size_t i = 0; i < 256; i++)
+    decodeHexTable[i] = invalid;
```
I would keep the `std::array<byte, 256>` here. If CodeQL is complaining, it means that there's a bug in the query that I wrote.

neheb replied:
So the other location uses std::vector (heap vs stack allocation) to work around compiler bugs. See https://github.com/Exiv2/exiv2/commit/c8fab3abad2ec4637b8670fd2c67d09c8f6e9f04

kevinbackhouse replied:
Ok

neheb replied:
On second thought, it's easier to use fill() than a custom loop. Backing out of this.

kevinbackhouse commented:
The `decodeHexTable` change looks unnecessary to me. Other than that, everything else is good.

neheb committed:
clang: fix warnings

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
use more stringFormat

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
clang-analyzer: fix NewDelete error

add calls doAdd which calls push_back and release(). It's more reliable
to call back() in this case as the formerly unique_ptr is at the end.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb commented:
@kevinbackhouse I was right to question this get() usage. clang-analyzer complains with:

```
../src/crwimage_int.cpp:577:3: warning: Use of memory after it is freed [clang-analyzer-cplusplus.NewDelete]
  577 |   return cc;
      |   ^
../src/crwimage_int.cpp:992:7: note: Assuming the condition is true
  992 |   if (!buf.empty()) {
      |       ^~~~~~~~~~~~
../src/crwimage_int.cpp:992:3: note: Taking true branch
  992 |   if (!buf.empty()) {
      |   ^
../src/crwimage_int.cpp:993:5: note: Calling 'CiffHeader::add'
  993 |     pHead.add(pCrwMapping.crwTagId_, pCrwMapping.crwDir_, std::move(buf));
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/crwimage_int.cpp:516:3: note: Taking false branch
  516 |   if (!pRootDir_) {
      |   ^
../src/crwimage_int.cpp:519:20: note: Calling 'CiffComponent::add'
  519 |   if (auto child = pRootDir_->add(crwDirs, crwTagId)) {
      |                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/crwimage_int.cpp:525:10: note: Calling 'CiffDirectory::doAdd'
  525 |   return doAdd(crwDirs, crwTagId);
      |          ^~~~~~~~~~~~~~~~~~~~~~~~
../src/crwimage_int.cpp:546:7: note: Assuming the condition is false
  546 |   if (!crwDirs.empty()) {
      |       ^~~~~~~~~~~~~~~~
../src/crwimage_int.cpp:546:3: note: Taking false branch
  546 |   if (!crwDirs.empty()) {
      |   ^
../src/crwimage_int.cpp:571:8: note: 'cc' is null
  571 |   if (!cc) {
      |        ^~
../src/crwimage_int.cpp:571:3: note: Taking true branch
  571 |   if (!cc) {
      |   ^
../src/crwimage_int.cpp:573:14: note: Calling 'make_unique<Exiv2::Internal::CiffEntry, unsigned short &, unsigned short>'
  573 |     auto m = std::make_unique<CiffEntry>(crwTagId, tag());
      |              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/unique_ptr.h:962:30: note: Memory is allocated
  962 |     { return unique_ptr<_Tp>(new _Tp(std::forward<_Args>(__args)...)); }
      |                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/crwimage_int.cpp:573:14: note: Returned allocated memory
  573 |     auto m = std::make_unique<CiffEntry>(crwTagId, tag());
      |              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/crwimage_int.cpp:575:5: note: Calling 'CiffComponent::add'
  575 |     add(std::move(m));
      |     ^~~~~~~~~~~~~~~~~
../src/crwimage_int.cpp:144:28: note: Calling '~unique_ptr'
  144 |   doAdd(std::move(component));
      |                            ^
/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/unique_ptr.h:360:2: note: Taking true branch
  360 |         if (__ptr != nullptr)
      |         ^
/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/unique_ptr.h:361:4: note: Calling 'default_delete::operator()'
  361 |           get_deleter()(std::move(__ptr));
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/unique_ptr.h:85:2: note: Memory is released
   85 |         delete __ptr;
      |         ^~~~~~~~~~~~
/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/unique_ptr.h:361:4: note: Returning; memory was released via 2nd parameter
  361 |           get_deleter()(std::move(__ptr));
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/crwimage_int.cpp:144:28: note: Returning from '~unique_ptr'
  144 |   doAdd(std::move(component));
      |                            ^
../src/crwimage_int.cpp:575:5: note: Returning; memory was released
  575 |     add(std::move(m));
      |     ^~~~~~~~~~~~~~~~~
../src/crwimage_int.cpp:577:3: note: Use of memory after it is freed
  577 |   return cc;
```

add() eventually calls release() on the pointer which also ends up freeing it. so m.get() returns garbage. Supposedly.

kevinbackhouse commented:
> @kevinbackhouse I was right to question this get() usage. clang-analyzer complains with:

@neheb: which log file is this in? Or can you give me instructions to reproduce it locally?

neheb commented:
> > @kevinbackhouse I was right to question this get() usage. clang-analyzer complains with:
> 
> @neheb: which log file is this in? Or can you give me instructions to reproduce it locally?

assuming your cmake configuration includes -DCMAKE_EXPORT_COMPILE_COMMANDS=ON , run

```
run-clang-tidy -header-filter=../ -checks=clang-analyzer\* ../src/crwimage_int.cpp
```

neheb committed:
use some std::min

Signed-off-by: Rosen Penev <rosenp@gmail.com>

kevinbackhouse commented:
> > > @kevinbackhouse I was right to question this get() usage. clang-analyzer complains with:
> > 
> > 
> > @neheb: which log file is this in? Or can you give me instructions to reproduce it locally?
> 
> assuming your cmake configuration includes -DCMAKE_EXPORT_COMPILE_COMMANDS=ON , run
> 
> ```
> run-clang-tidy -header-filter=../ -checks=clang-analyzer\* ../src/crwimage_int.cpp
> ```

@neheb: you already fixed it in https://github.com/Exiv2/exiv2/pull/3136/commits/f81fd334fa4c6073c27a91eefd77e5dd540b6755, right?

neheb commented:
Yes

kevinbackhouse reviewed (approved):
I think this looks good now. In future though, please could you not bundle so many unrelated changes into one PR? I find smaller PRs easier to review. Also, if a change causes a regression, I find it's easier to track down what caused it if each individual PR is smaller.
